File: DeskFunUI/DeskFunUI/DABezelView.m
//
//  DABezelView.m
//  DAGame
//
//  Created by Christoph Leimbrock on 25/10/13.
//  Copyright (c) 2013 CCL. All rights reserved.
//

#import "DABezelView.h"
#import "DAWindowTheme.h"

@implementation DABezelView

- (id)initWithFrame:(NSRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        self.bezelWidth = 2.0;
    }
    return self;
}

- (id)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    if(self)
    {
        self.bezelWidth = 2.0;
    }
    return self;
}


- (id)init
{
    self = [super init];
    if(self)
    {
        self.bezelWidth = 2.0;
    }
    return self;
}

- (BOOL)isOpaque
{
    return [[[DAWindowTheme sharedTheme] backgroundColor] alphaComponent] == 1.0;;
}

- (void)drawRect:(NSRect)dirtyRect
{
    DAWindowTheme *theme = [DAWindowTheme sharedTheme];
    CGFloat bezelWidth = self.bezelWidth;

    NSRect bounds = self.bounds;
    CGFloat maxY = NSMaxY(bounds);
    CGFloat maxX = NSMaxX(bounds);

    NSBezierPath *path = [NSBezierPath bezierPath];
    [path moveToPoint:NSMakePoint(0, 0)];
    [path lineToPoint:NSMakePoint(0, maxY)];
    [path lineToPoint:NSMakePoint(maxX, maxY)];

    [path lineToPoint:NSMakePoint(maxX-bezelWidth, maxY-bezelWidth)];
    [path lineToPoint:NSMakePoint(bezelWidth, maxY-bezelWidth)];
    [path lineToPoint:NSMakePoint(bezelWidth, bezelWidth)];
    [path lineToPoint:NSMakePoint(0, 0)];

    [self.swapColors ? [theme bezelColorLight] : [theme bezelColorDark] set];
    [path fill];

    path = [NSBezierPath bezierPath];
    [path moveToPoint:NSMakePoint(0, 0)];
    [path lineToPoint:NSMakePoint(maxX, 0)];
    [path lineToPoint:NSMakePoint(maxX, maxY)];

    [path lineToPoint:NSMakePoint(maxX-bezelWidth, maxY-bezelWidth)];
    [path lineToPoint:NSMakePoint(maxX-bezelWidth, bezelWidth)];
    [path lineToPoint:NSMakePoint(bezelWidth, bezelWidth)];
    [path lineToPoint:NSMakePoint(0, 0)];

    [self.swapColors ? [theme bezelColorDark] : [theme bezelColorLight] set];
    [path fill];

    [self.backgroundColor ?: [theme backgroundColor] setFill];
    NSRectFill(NSInsetRect(self.bounds, bezelWidth, bezelWidth));
}

@end

File: DeskFunUI/DeskFunTouchUI/DABezelView.swift
//
//  DABezelView.swift
//  DeskFunUI
//
//  Copyright © 2015 ccl. All rights reserved.
//

import Foundation
import UIKit

public class DABezelView : UIView {
    public var bezelWidth: CGFloat = 2.0;
    public var invertColors: Bool = false;
    public var background: UIColor = DAWindowTheme.BackgroundColor;

    override public func drawRect(rect: CGRect) {
        let bounds = self.bounds;
        let maxX = CGRectGetMaxX(bounds);
        let maxY = CGRectGetMaxY(bounds);

        let innerColor = invertColors ? DAWindowTheme.BezelLightColor : DAWindowTheme.BezelDarkColor;
        let outerColor = invertColors ? DAWindowTheme.BezelDarkColor : DAWindowTheme.BezelLightColor;

        var path = UIBezierPath();
        path.moveToPoint(CGPointMake(0, maxY));
        path.addLineToPoint(CGPointMake(0, 0));
        path.addLineToPoint(CGPointMake(maxX,0));
        path.addLineToPoint(CGPointMake(maxX-bezelWidth, bezelWidth));
        path.addLineToPoint(CGPointMake(bezelWidth, bezelWidth));
        path.addLineToPoint(CGPointMake(bezelWidth, maxY-bezelWidth));
        path.addLineToPoint(CGPointMake(0, maxY));
        innerColor.setFill();
        path.fill();

        path = UIBezierPath();
        path.moveToPoint(CGPointMake(0, maxY));
        path.addLineToPoint(CGPointMake(maxX,maxY));
        path.addLineToPoint(CGPointMake(maxX,0));
        path.addLineToPoint(CGPointMake(maxX-bezelWidth,bezelWidth));
        path.addLineToPoint(CGPointMake(maxX-bezelWidth,maxY-bezelWidth));
        path.addLineToPoint(CGPointMake(bezelWidth,maxY-bezelWidth));
        path.addLineToPoint(CGPointMake(0, maxY));

        outerColor.setFill();
        path.fill();

        background.setFill();
        UIRectFill(CGRectInset(bounds, bezelWidth, bezelWidth));
    }
}